std/haxe/iterators/StringIteratorUnicode.hx

```diff
@@ -31,28 +31,47 @@ package haxe.iterators;
 	String's length due to the fact that it deals with surrogate pairs.
 **/
 class StringIteratorUnicode {
+	#if js
+	var native_iter: Dynamic;
+	var done = false;
+	#else
 	var offset = 0;
+	#end
+
 	var s:String;
 
 	/**
 		Create a new `StringIteratorUnicode` over String `s`.
 	**/
 	public inline function new(s:String) {
 		this.s = s;
+		#if js
+		this.native_iter = untyped __js__("{0}[Symbol.iterator]()", this.s);
+		#end
 	}
 
 	/**
 		See `Iterator.hasNext`
 	**/
 	public inline function hasNext() {
+		#if js
+		return !this.done;
+		#else
 		return offset < s.length;
+		#end
 	}
 
 	/**
 		See `Iterator.next`
 	**/
 	@:access(StringTools)
 	public inline function next() {
+		#if js
+		var next = untyped __js__("{0}.next()", this.native_iter);
```

Comment:
> The `[@@iterator]()` method returns a new `Iterator` object that iterates over the code points of a String value, returning each code point as a String value.

You will have to do `charCodeAt(0)` to return the actual codepoints for BMP characters.

For non-BMP characters, e.g. `🐄` (U+1F404), you will need to combine the surrogates, see the `#if utf16` part in the original code. Checking if the first code point is `>= MIN_SURROGATE_CODE_POINT` should be enough to know that you need to combine it with the second code point.

Reply:
thanks. I'll use native `codePointAt`, instead of `charCodeAt` with checking surrogate code points. This should work well also for non-BMP characters.